src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/Symbols/EEMethodSymbol.vb

```diff
@@ -513,11 +519,17 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator
             Dim displayClassVariables = PooledDictionary(Of String, DisplayClassVariable).GetInstance()
             For Each pair In _displayClassVariables
                 Dim variable = pair.Value
-                Dim displayClassInstanceFromLocal = TryCast(variable.DisplayClassInstance, DisplayClassInstanceFromLocal)
-                Dim displayClassInstance = If(displayClassInstanceFromLocal Is Nothing,
-                        DirectCast(New DisplayClassInstanceFromMe(Me.Parameters(0)), DisplayClassInstance),
-                        New DisplayClassInstanceFromLocal(DirectCast(localMap(displayClassInstanceFromLocal.Local), EELocalSymbol)))
-                variable = New DisplayClassVariable(variable.Name, variable.Kind, displayClassInstance, variable.DisplayClassFields)
+                Dim oldDisplayClassInstance = variable.DisplayClassInstance
+
+                ' Note: even though ToOtherMethod is available on oldDisplayClassInstance,
+                ' we can't call it because that would produce a new LocalSymbol that would
+                ' not be ReferenceEquals to the one in this.LocalsForBinding.
+                Dim oldDisplayClassInstanceFromLocal = TryCast(oldDisplayClassInstance, DisplayClassInstanceFromLocal)
+                Dim newDisplayClassInstance = If(oldDisplayClassInstanceFromLocal Is Nothing,
+                    oldDisplayClassInstance.ToOtherMethod(Me, Me.TypeMap),
+                    New DisplayClassInstanceFromLocal(DirectCast(localMap(oldDisplayClassInstanceFromLocal.Local), EELocalSymbol)))
+
+                variable = variable.SubstituteFields(newDisplayClassInstance, Me.TypeMap)
```

Comment:
Can this be extracted as a method and shared with the similar code in the constructor?